sdlib/d/sync/sharedlock.d

```diff
@@ -0,0 +1,421 @@
+		count = 0;
+	}
+}
+
+unittest sharedLocks {
+	// Simple testing of shared locks, then test multiple threads acquiring
+	// the lock.
+	static runThread(void* delegate() dg) {
+		static struct Delegate {
+			void* ctx;
+			void* function(void*) fun;
+		}
+
+		auto x = *(cast(Delegate*) &dg);
+
+		import core.stdc.pthread;
+		pthread_t tid;
+		auto r = pthread_create(&tid, null, x.fun, x.ctx);
+		assert(r == 0, "Failed to create thread!");
+
+		return tid;
+	}
+
+	shared SharedLock lock;
+
+	lock.sharedLock();
+	assert(lock.count == 1);
+	assert(!lock.mutex.isHeld());
+
+	lock.sharedLock();
+	assert(lock.count == 2);
+
+	lock.sharedUnlock();
+	assert(lock.count == 1);
+
+	import d.sync.atomic;
+	shared Mutex mutex;
+	shared Atomic!uint numLocked;
+	bool exitThread;
+
+	void* run() {
+		lock.sharedLock();
+		scope(exit) lock.sharedUnlock();
+
+		numLocked.fetchAdd(1);
+
+		mutex.lock();
+		scope(exit) mutex.unlock();
+
+		bool shouldExit() {
+			return exitThread;
+		}
+
+		mutex.waitFor(shouldExit);
+		numLocked.fetchSub(1);
+		return null;
+	}
+
+	import core.stdc.pthread;
+	pthread_t[50] tids;
+	foreach (i; 0 .. tids.length) {
+		tids[i] = runThread(run);
+	}
+
+	{
+		mutex.lock();
+		scope(exit) mutex.unlock();
+
+		bool allThreadsReady() {
+			return numLocked.load() == tids.length;
+		}
+
+		mutex.waitFor(allThreadsReady);
+		assert(lock.count == 51);
+
+		exitThread = true;
+
+		bool allThreadsDone() {
+			return numLocked.load() == 0;
+		}
+
+		mutex.waitFor(allThreadsDone);
+		assert(lock.count == 1);
+	}
+
+	void* ret;
+	foreach (i; 0 .. tids.length) {
+		pthread_join(tids[i], &ret);
+	}
+
+	lock.sharedUnlock();
+	assert(lock.count == 0);
+}
+
+unittest exclusiveLock {
+	// Test using the exclusive lock. First, check that the state is as
+	// expected, then test trying to lock exclusively while holding a
+	// shared lock, and finally test exclusive locks as a mutex between 2
+	// threads.
+	static runThread(void* delegate() dg) {
+		static struct Delegate {
+			void* ctx;
+			void* function(void*) fun;
+		}
+
+		auto x = *(cast(Delegate*) &dg);
+
+		import core.stdc.pthread;
+		pthread_t tid;
+		auto r = pthread_create(&tid, null, x.fun, x.ctx);
+		assert(r == 0, "Failed to create thread!");
+
+		return tid;
+	}
+
+	shared SharedLock lock;
+	lock.exclusiveLock();
+	assert(lock.count == SharedLock.Exclusive);
+
+	lock.exclusiveUnlock();
+	assert(lock.count == 0);
+
+	lock.sharedLock();
+	assert(lock.count == 1);
+
+	// Attempt to exclusive lock while the shared lock is held.
+	import core.stdc.pthread;
+	import d.sync.atomic;
+	shared Atomic!uint state;
+
+	void* takeExclusiveLock() {
+		state.fetchAdd(1);
+		lock.exclusiveLock();
+		state.fetchAdd(1);
+		return cast(void*) cast(size_t) lock.count;
+	}
+
+	void* takeSharedLock() {
+		lock.sharedLock();
+		state.fetchAdd(1);
+		return null;
+	}
+
+	auto t1 = runThread(takeExclusiveLock);
+
+	{
+		bool triedExclusiveLock() {
+			return lock.count >= SharedLock.Exclusive;
+		}
```

Comment:
You are trying to test the behavior of the lock, so you should be waiting on state and not on the mutex internals.

Note that here, the count will go to SharedLock.Exclusive when a reader unlocks and a writer is pending, so you have no idea that the writer actually took the lock at that point.

Reply:
I can't atomically set the exclusive lock, and increment some counter to check. Basically, the thread will increment the external counter and then at some point after flag that it is seeking the exclusive lock. I need to wait on that internal. I thought using the internal mutex to wait was a decent approach because it flags exactly when the exclusive thread goes to sleep waiting for the count to reach 0.

I'm not sure how else to test this without checking that `lock.count` internal.

Reply:
This is slightly different now, I still don't know how to wait for this.